### ISSUE TITLE
Support tag filtering alongside branch filtering

### ISSUE THREAD

**Issue by alecwcp** [CircleCI-Public/slack-orb#187]
## Describe Request:
The ability to filter by branch is really useful - we use this to filter notifications to only be sent when a deployment to one of our staging sites (controlled by branch name) fails, and also have a separate notification for failed develop branch builds.

Our production deploys are done on a semver tag. CircleCI supports filtering the jobs themselves based on the tag or the branch, but this orb currently only supports branch filtering. Adding tag filtering too would be a useful addition to allow us to notify on these with a separate message / to a separate channel / etc.
## Examples:
Since `branch_pattern` is used for branches, a new parameter of `tag_pattern` that accepts a regex might be fitting. It would work as an OR condition with the `branch_pattern` - so one notification section could be used for both tagged builds and branch builds.
## Supporting Documentation Links:
- `branch_pattern` parameter documentation here https://circleci.com/developer/orbs/orb/circleci/slack
-  Filtering jobs by branch/tag documentation here https://circleci.com/docs/2.0/workflows/#using-regular-expressions-to-filter-tags-and-branches

**Comment by KyleTryon** [comment]
Hello,

Does the following use-case solve your need?

https://circleci.com/developer/orbs/orb/circleci/slack#usage-successful_tagged_deployment

```
version: '2.1'
orbs:
  node: 'circleci/node:4.0'
  slack: circleci/slack@4.0
jobs:
  deploy:
    executor:
      name: node/default
    steps:
      - checkout
      - node/install-packages
      - run:
          command: npm run deploy
      - slack/notify:
          channel: ABCXYZ
          event: success
          template: basic_fail_1
workflows:
  deploy_and_notify:
    jobs:
      - deploy:
          context:
            - slack-secrets
```

Workflow filters for instance already support tagging, so that a job will only run in the event the job has been tagged, at which point, no special parameter for the orb will be needed.

**Comment by alecwcp** [comment]
Quick response :slightly_smiling_face: 

Currently the only difference between the `deploy_prod` and `deploy_nonprod` jobs is the context. So to avoid duplication in our config file we define the deploy job once then use yaml anchors to use it for both the `deploy_prod` and `deploy_nonprod` jobs with just the context difference.

I think we could pull some of that out, it would create some duplication in the file but would accomplish what you suggest above.

However, an actual deployment includes running the automated tests prior to the deploy too (separate jobs) - if one of those fails and it is part of a workflow that would result in a deployment from a tag we'd also want to be able to send a slack notification that a live deployment failed - but the test job is the same as runs everywhere (on PRs, on nonprod branch deployments, etc.) so jsut having the slack/notify on that job would trigger slack notifications every time someone made a PR that failed tests.

**Comment by KyleTryon** [comment]
> Currently the only difference between the `deploy_prod` and `deploy_nonprod` jobs is the context. So to avoid duplication in our config file we define the deploy job once then use yaml anchors to use it for both the `deploy_prod` and `deploy_nonprod` jobs with just the context difference.

If you have defined a job in your config, you can simply call it twice, you can give it a custom name parameter too if desired. In the same way you have your `test` and `deploy` jobs here, you could call those same jobs twice within the same workflow even, but they will need unique `name` parameters.

Ah, ok I see your point here. Perhaps a tag parameter could be used for this reason. Allow me to share how this could be accomplished now, but I think we can open this conversation up for adding that parameter given the use case.


Example:

```
version: '2.1'
orbs:
  node: 'circleci/node:4.0'
  slack: circleci/slack@4.0
jobs:
  deploy:
    executor:
      name: node/default
    steps:
      - checkout
      - node/install-packages
      - run:
          command: npm run deploy
      - slack/notify:
          channel: ABCXYZ
          event: fail
          template: basic_fail_1
      - slack/notify:
          channel: ABCXYZ
          event: success
          template: successful_deployment
  test:
    executor:
      name: node/default
    steps:
      - checkout
      - node/install-packages
      - run:
          command: npm run tests
      - slack/notify:
          # How to make this only trigger if running for a tag that matched a regex pattern
          channel: ABCXYZ
          event: fail
          template: basic_fail_1
workflows:
test_and_deploy_not_tagged:
    jobs:
      - test:
          context:
            - slack-secrets
      - deploy:
          requires:
            - test
          filters:
              branches:
                 only:
                   - staging
          context:
            - slack-secrets
  test_and_deploy_tagged:
    jobs:
      - test:
           tags:
              only:
                - /[0-9]+(\.[0-9]+)*/
          context:
            - slack-secrets
      - deploy:
          requires:
            - test
          filters:
              branches:
                 only:
                   - staging
          context:
            - slack-secrets
```

**Comment by KyleTryon** [comment]
Resolved in #196 
Thank you!
